src/main/java/io/github/bobocodebreskul/context/scan/RecursiveClassPathAnnotatedBeanScanner.java

```diff
@@ -0,0 +1,78 @@
+package io.github.bobocodebreskul.context.scan;
+
+import static io.github.bobocodebreskul.context.support.GeneralConstants.EMPTY;
+import static java.util.Arrays.asList;
+import static java.util.function.Predicate.not;
+import static java.util.stream.Collectors.toSet;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * Implementation of {@link ClassPathAnnotatedBeanScanner} to recursively find all bean definitions
+ * in package. Additionally, performs package scan of all found configurations.
+ *
+ * @author Vitalii Katkov
+ */
+public class RecursiveClassPathAnnotatedBeanScanner implements ClassPathAnnotatedBeanScanner {
+
+  private static final String PACKAGE_DELIMITER = ".";
+
+  @Override
+  public Set<String> scan(String... scanPackages) {
+    Set<String> allFoundClasses = new HashSet<>();
+
+    Queue<String> remainingScanPackages = new ArrayDeque<>(asList(scanPackages));
+    Set<String> processedScanPackages = new HashSet<>();
+    while (!remainingScanPackages.isEmpty()) {
+      String scanPackage = remainingScanPackages.poll();
+      processedScanPackages.add(scanPackage);
+
+      Set<String> foundClasses = scanSingle(scanPackage);
+      processedScanPackages.addAll(collectScannedPackages(foundClasses));
+      allFoundClasses.addAll(foundClasses);
+
+      foundClasses.stream()
+          .filter(this::isConfigurationClass)
+          .map(this::findConfigurationScanPackages)
+          .filter(configurationScanPackages -> configurationScanPackages.length > 0)
```

Comment:
Do we need this filter?

Reply:
Just for not to call Arrays.stream on empty array, little optimization.